### PR TITLE
Add row indexing to filter docstring and examples.

### PR DESCRIPTION
**Pull request by nathanrboyer**
It was shown to me recently that column selectors can be used inside filter function definitions, which is incredibly useful. I want to add a brief statement and example to the `filter` and `filter!` documentation to show that usage.

### CODE REVIEW

**Comment by nathanrboyer** [comment]
This is the functionality I want to illustrate:
```julia
julia> df = DataFrame(node=1:4, x=[1,0,9,5], y=[0,0,12,8])
4×3 DataFrame
 Row │ node   x      y     
     │ Int64  Int64  Int64 
─────┼─────────────────────
   1 │     1      1      0
   2 │     2      0      0
   3 │     3      9     12
   4 │     4      5      8

julia> filter(row -> !all(x -> x == 0, row[Not(:node)]), df)
3×3 DataFrame
 Row │ node   x      y     
     │ Int64  Int64  Int64
─────┼─────────────────────
   1 │     1      1      0
   2 │     3      9     12
   3 │     4      5      8
```

However, both of these examples are more complicated than I intended:
```julia
julia> df = DataFrame(x = [3, 1, 2, 1], y = ["b", "c", "a", "b"])
4×2 DataFrame
 Row │ x      y      
     │ Int64  String 
─────┼───────────────
   1 │     3  b
   2 │     1  c
   3 │     2  a
   4 │     1  b

julia> filter(row -> row[Not(:y)]... > 1, df)
2×2 DataFrame
 Row │ x      y      
     │ Int64  String
─────┼───────────────
   1 │     3  b
   2 │     2  a

julia> filter(row -> only(row[Not(:y)]) > 1, df)
2×2 DataFrame
 Row │ x      y      
     │ Int64  String
─────┼───────────────
   1 │     3  b
   2 │     2  a
```

Any better ideas?

**Comment by nathanrboyer** [comment]
Also, I think the row indexing rules need updated in the Indexing API documentation.

> The rules for a valid type of index into a row are the following:
>    * a value, later denoted as row:
>      * an Integer that is not Bool;

`Symbol`s and `String`s also work:
```julia
julia> filter(row -> row[:x] > 1, df)
2×2 DataFrame
 Row │ x      y      
     │ Int64  String
─────┼───────────────
   1 │     3  b
   2 │     2  a

julia> filter(row -> row["x"] > 1, df)
2×2 DataFrame
 Row │ x      y      
     │ Int64  String
─────┼───────────────
   1 │     3  b
   2 │     2  a
```

**Comment by bkamins** [comment]
> Also, I think the row indexing rules need updated in the Indexing API documentation.

These are rules for row index (which is not available in `DataFrameRow`). `DataFrameRow` accepts only column index, which is defined as:
> a value, later denoted as col:
> * a Symbol;
> * an AbstractString;
> * an Integer that is not Bool;

File: docs/src/lib/indexing.md
Comment (bkamins) [inline docs/src/lib/indexing.md:54]: rows cannot be indexed with `Symbol` nor `AbstractString`
```suggestion
```

File: docs/src/lib/indexing.md
Comment (bkamins) [inline docs/src/lib/indexing.md:51]: this probably disambiguates the issue
```suggestion
The rules for a valid type of row index are the following:
```

**Comment by bkamins** [comment]
> Any better ideas, or is the first example okay?

Your examples are valid but are complex (i.e. they use multiple elements that the reader must know to understand them). Do you think that just showing e.g.:

```
filter(row -> row[:x] > 1, df)
```
would achieve the objective you had?

**Comment by nathanrboyer** [comment]
> These are rules for row index (which is not available in DataFrameRow).

The "into" language was confusing. I would index into a row with a column. Your fix is good.

> Do you think that just showing e.g. `filter(row -> row[:x] > 1, df)` would achieve the objective you had?

Yes, let's just go with that. It implies that the more complex options are possible.

File: docs/src/lib/indexing.md
Comment (nathanrboyer) [inline docs/src/lib/indexing.md:51]: Maybe delete the word "types" since I'm not sure `:` and `!` count as types. I also considered "The rules for a valid row index are the following:" instead.

Comment (bkamins) [inline docs/src/lib/indexing.md:51]:
```suggestion
The following values are a valid row index:
```

Comment (bkamins) [reply docs/src/lib/indexing.md:51]: we index with values not types.

File: docs/src/lib/indexing.md
Comment (bkamins) [inline docs/src/lib/indexing.md:59]:
```suggestion
      [InvertedIndices.jl](https://github.com/JuliaData/InvertedIndices.jl));
```

File: docs/src/lib/indexing.md
Comment (bkamins) [inline docs/src/lib/indexing.md:60]:
```suggestion
    * a literal colon `:` (selects all columns with copying);
```

File: docs/src/lib/indexing.md
Comment (bkamins) [inline docs/src/lib/indexing.md:18]:
```suggestion
The following values are a valid column index:
```

File: src/abstractdataframe/abstractdataframe.jl
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:996]:
```suggestion
with column selectors inside the definition of `fun`.)
```

File: src/abstractdataframe/abstractdataframe.jl
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:1126]:
```suggestion
with column selectors inside the definition of `fun`.)
```

File: docs/src/lib/indexing.md
Comment (nathanrboyer) [reply docs/src/lib/indexing.md]: Is a vector a value? I just deleted both "types" and "values". I don't think either word helps clarify. The list contains mostly types, but I think it is clear that `row[Int32]` wouldn't make any sense.

Comment (bkamins) [reply docs/src/lib/indexing.md]: vector is a value. Types are e.g. `Int64` or `Vector{Int}`, but you are not allowed to use them for indexing.

File: docs/src/lib/indexing.md
Comment (nathanrboyer) [inline docs/src/lib/indexing.md:31]: Any way to make this link more future-proof?

Comment (bkamins) [reply docs/src/lib/indexing.md:31]: I think it is OK

File: docs/src/lib/indexing.md
Comment (nathanrboyer) [reply docs/src/lib/indexing.md]: Okay, I put "value" back in line 51, but then "value" in line 52 must change. I used "scalar" in line 52 in place of "value". I'm not sure that this is any clearer, but I'll let you choose your preference: see commit 29c2412.

File: docs/src/lib/indexing.md
Comment (bkamins) [inline docs/src/lib/indexing.md:61]:
```suggestion
    * a literal exclamation mark `!` (selects all rows without copying).
```

**Review by bkamins — APPROVED** [review]
Looks good. Thank you! @nalimilan - can you please double check the language?

File: docs/src/lib/indexing.md
Comment (nalimilan) [inline docs/src/lib/indexing.md:32]:
```suggestion
    * a [regular expression](https://docs.julialang.org/en/v1/manual/strings/#Regular-Expressions)
      (will be expanded to a vector of matching column names);
```

File: src/abstractdataframe/abstractdataframe.jl
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:996]: Putting text in parentheses is weird. If we think this text is useful here it shouldn't hide it.
```suggestion
In that case, individual elements may be accessed by indexing the passed `DataFrameRow`s
with column selectors inside `fun`.
```

File: src/abstractdataframe/abstractdataframe.jl
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:1126]:
```suggestion
In that case, individual elements may be accessed by indexing the passed `DataFrameRow`s
with column selectors inside `fun`.
```

File: src/abstractdataframe/abstractdataframe.jl
Comment (nathanrboyer) [reply src/abstractdataframe/abstractdataframe.jl:996]: I should probably mention the `row.x` possibility here too. How about the following?
"If `cols` is not specified then the predicate `fun` is passed `DataFrameRow`s.
Elements of a `DataFrameRow` may be accessed with dot syntax or column indexing inside `fun`."

Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:996]: OK with me

**Comment by nathanrboyer** [comment]
@nalimilan I'll take your opinion on wording surrounding commit 29c2412 too.

File: src/abstractdataframe/abstractdataframe.jl
Comment (nalimilan) [reply src/abstractdataframe/abstractdataframe.jl:996]: OK.

**Comment by nathanrboyer** [comment]
Should be ready to merge now.

**Comment by bkamins** [comment]
Looks good. I will merge once @nalimilan approves.

**Review by nalimilan — APPROVED** [review]

**Comment by nalimilan** [comment]
Thanks!

**Pull request merged**
